### PR TITLE
Use a list for Azure services in reconcilers

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**: Since we are reconciling (and deleting in reverse order) Azure services one by one in reconcilers, this is a small optimization to move the reconcile and delete to a list and range through the services instead of making a Reconcile/Delete call for each. This is possible now that we've refactored all services to implement `Azure.Reconciler`. Also adds unit tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Use a list for Azure services in AzureCluster and AzureMachine reconcilers
```

### CODE REVIEW

File: controllers/azurecluster_reconciler.go [hunk 1]
```diff
@@ -63,22 +54,23 @@ func newAzureClusterService(scope *scope.ClusterScope) (*azureClusterService, er
...
-		publicIPSvc:      publicips.New(scope),
-		loadBalancerSvc:  loadbalancers.New(scope),
-		privateDNSSvc:    privatedns.New(scope),
-		bastionSvc:       bastionhosts.New(scope),
-		skuCache:         skuCache,
-		peeringsSvc:      vnetpeerings.New(scope),
-		tagsSvc:          tags.New(scope),
+		scope: scope,
+		services: []azure.Reconciler{
+			groups.New(scope),
```
Comment (CecileRobertMichon) [inline controllers/azurecluster_reconciler.go:60]: this is the same order as the Reconcile() calls below

File: controllers/azuremachine_reconciler.go [hunk 2]
```diff
@@ -60,17 +54,19 @@ func newAzureMachineService(machineScope *scope.MachineScope) (*azureMachineServ
...
-		vmExtensionsSvc:      vmextensions.New(machineScope),
-		availabilitySetsSvc:  availabilitysets.New(machineScope, cache),
-		skuCache:             cache,
+		scope: machineScope,
+		services: []azure.Reconciler{
+			publicips.New(machineScope),
+			inboundnatrules.New(machineScope),
+			networkinterfaces.New(machineScope, cache),
+			availabilitysets.New(machineScope, cache),
+			disks.New(machineScope),
```
Comment (CecileRobertMichon) [inline controllers/azuremachine_reconciler.go:63]: this is the same order as Reconcile() below except for disks that were not present before in Reconcile (it's a no-op) but needs to be deleted _after_ the VM (hence needs to be before VMs in the list)

**Review by devigned — APPROVED** [review]
I love this changeset! This improves the readability of the reconciler so much. 

/lgtm

**Review by mboersma — APPROVED** [review]
/lgtm

**Review by jackfrancis — APPROVED** [review]
/lgtm

/assign @shysank

File: controllers/azurecluster_reconciler.go [hunk 3]
```diff
@@ -152,46 +102,14 @@ func (s *azureClusterService) Delete(ctx context.Context) error {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "controllers.azureClusterService.Delete")
 	defer done()
 
-	if err := s.groupsSvc.Delete(ctx); err != nil {
+	if err := s.services[0].Delete(ctx); err != nil {
```
Comment (Jont828) [inline controllers/azurecluster_reconciler.go:105]: Could we set a variable like `groupSvc` before placing it in the list so we call that explicitly here?

File: controllers/azurecluster_reconciler.go [hunk 4]
```diff
@@ -152,46 +102,14 @@ func (s *azureClusterService) Delete(ctx context.Context) error {
...
-				return errors.Wrap(err, "failed to delete network security group")
-			}
-
-			if err := s.vnetSvc.Delete(ctx); err != nil {
-				return errors.Wrap(err, "failed to delete virtual network")
+			// If the resource group is not managed we need to delete resources inside the group one by one.
+			// services are deleted in reverse order from the order in which they are reconciled.
+			for i := len(s.services) - 1; i >= 1; i-- {
+				if err := s.services[i].Delete(ctx); err != nil {
+					return errors.Wrap(err, "failed to delete AzureCluster service")
```
Comment (Jont828) [inline controllers/azurecluster_reconciler.go:111]: Can we access the `serviceName` variable from `s.services[i]`? That way we could do something like `errors.Wrap(err, fmt.Sprintf("failed to delete AzureCluster service '%s'", serviceName)`.

File: controllers/azurecluster_reconciler_test.go [hunk 5]
```diff
@@ -30,72 +30,107 @@ import (
 	"sigs.k8s.io/cluster-api-provider-azure/azure/scope"
 	"sigs.k8s.io/cluster-api-provider-azure/azure/services/resourceskus"
 	gomockinternal "sigs.k8s.io/cluster-api-provider-azure/internal/test/matchers/gomock"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
 )
 
-type expect func(grp *mock_azure.MockReconcilerMockRecorder, vnet *mock_azure.MockReconcilerMockRecorder, sg *mock_azure.MockReconcilerMockRecorder, rt *mock_azure.MockReconcilerMockRecorder, sn *mock_azure.MockReconcilerMockRecorder, natg *mock_azure.MockReconcilerMockRecorder, pip *mock_azure.MockReconcilerMockRecorder, lb *mock_azure.MockReconcilerMockRecorder, dns *mock_azure.MockReconcilerMockRecorder, bastion *mock_azure.MockReconcilerMockRecorder, peer *mock_azure.MockReconcilerMockRecorder)
```
Comment (Jont828) [inline controllers/azurecluster_reconciler_test.go:35]: I'm glad we don't need this giant block anymore!

File: controllers/azurecluster_reconciler_test.go [hunk 6]
```diff
@@ -30,72 +30,107 @@ import (
...
+func TestAzureClusterServiceReconcile(t *testing.T) {
+	cases := map[string]struct {
+		expectedError string
+		expect        func(one *mock_azure.MockReconcilerMockRecorder, two *mock_azure.MockReconcilerMockRecorder, three *mock_azure.MockReconcilerMockRecorder)
```
Comment (Jont828) [inline controllers/azurecluster_reconciler_test.go:39]: Is there any significance in having three services in the test cases? Or is it just for the sake of convenience so it'll cover most cases?

File: controllers/azuremachine_reconciler.go (hunk 2 above)
Comment (Jont828) [reply controllers/azuremachine_reconciler.go:63]: What was our behavior before this change? Would the disk get deleted anyway?

Comment (CecileRobertMichon) [reply controllers/azuremachine_reconciler.go:63]: There was a disk delete call https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2146/files#diff-5a6d880b9a0be00329894fc2f52f15827dd6bdfe6b60aaec660728661023f41fL142 but no disk reconcile call. Disk reconcile is a no-op so adding it does not change behavior.

File: controllers/azurecluster_reconciler_test.go (hunk 6 above)
Comment (CecileRobertMichon) [reply controllers/azurecluster_reconciler_test.go:39]: no particular significance just chose a number big enough that it covers the basic edge cases (first, last, middle) but not too big that it makes the tests very long/repetitive and hard to read. 

Here since we're testing code with a for loop, the behavior should be the same whether there are 3, 4, 5, or 50 services. The goal is not to test the behavior of the for loop itself but more that for a given number of items our code is behaving as expected. I chose naming "one" "two" "three" since it makes the ordering aspect of the test very obvious and easy to follow (services are reconciled in order and deleted in reverse).

File: controllers/azurecluster_reconciler.go (hunk 3 above)
Comment (shysank) [reply controllers/azurecluster_reconciler.go:105]: Or can we make service registration a little bit more elaborate instead of just adding to an array. Something like
```
type RecocilerService struct {
 Name string,
 azure.Reconciler
}

func Register(serviceName, azure.Reconciler) {
}

```

This would help us easily add new apis to control reconciliation flow. For eg. for the case above, we can have a `GetService(name)` which iterates over the services list and returns the one with matching name. 
Or add a `Skip` option to the `ReconcilerService` if we want to skip reconciliation etc. wdyt?

Comment (CecileRobertMichon) [reply controllers/azurecluster_reconciler.go:105]: even if we assign it to a variable it wouldn't be accessible from here unless we add it as a separate field on the `azureClusterService` struct since we only have access to the `s` receiver from this func. we could assign a new variable here to make it explicit that it's what we expect but the variable would only be used once and it would still make the same assumption on the order so not sure that's an improvement, like this:

```
groupSvc := s.services[0]
if err := s.groupSvc.Delete(ctx); err != nil {
```

File: controllers/azurecluster_reconciler.go (hunk 4 above)
Comment (CecileRobertMichon) [reply controllers/azurecluster_reconciler.go:111]: not currently. Each service in `s.services` is a `azure.Reconciler` which only implements `Reconcile()` and `Delete()`. I actually thought about this as well and considered adding `ServiceName()` to the interface so we could do this but eventually decided not to pursue that route since it requires modifying every single service and the error we get back already contains the service name/type.

File: controllers/azurecluster_reconciler.go (hunk 3 above)
Comment (CecileRobertMichon) [reply controllers/azurecluster_reconciler.go:105]: @shysank I like the idea of registering the services. How would you set order and iterate over all the services without a list?

Comment (Jont828) [reply controllers/azurecluster_reconciler.go:105]: Could we make a pretty simple struct with an array and a map from a name to its element in the array? Something like
```
type ServiceWrapper struct {
    Services []azure.Reconciler 
    ServiceMap map[string]azure.Reconciler
}

func (s *ServiceWrapper) Add(name string, service azure.Reconciler) {
    s.Services = append(s.Services, service)
    s.ServiceMap[name] = service
}

func (s *ServiceWrapper) Get(name string) azure.Reconciler {
    return s.ServiceMap[name]
}
```

File: controllers/azurecluster_reconciler.go (hunk 4 above)
Comment (Jont828) [reply controllers/azurecluster_reconciler.go:111]: Where does the error add the service name? It looks like we add the `serviceName` in the errors coming out of `CreateResource()` and `DeleteResource()` in async.go, but if we just return an error like `errors.Errorf("%T is not a network.Subnet", result)` in the reconciler I'm not sure if that will include the service name.

File: controllers/azurecluster_reconciler.go (hunk 3 above)
Comment (shysank) [reply controllers/azurecluster_reconciler.go:105]: > How would you set order and iterate over all the services without a list?
There will still be some collection backing the services, just that it'd be something more than just []azure.Reonciler. Something like `[]ReconcilerService` (or servicesMap in @Jont828's example). If I want to add a new service, I'll just use the `Register` or `Add` api without having to care about the underlying collection. For ordering, we could start with the order in which the services were registered (add priorities later if needed). We can extend this further by tagging services from which we can extract groups. For example, there can be a group with services independent of each other, and services in that group can be reconciled concurrently while the group itself will be reconciled serially. (shameless plugin of a poc that I did earlier :) ).

Anyways, I'm fine with merging this PR with the current design since there has been so many reviews already, and we can discuss this in a separate issue.

Comment (CecileRobertMichon) [reply controllers/azurecluster_reconciler.go:105]: Let me try something out. Concurrency is out of scope and we should discuss it separately but I think I can at least get a service name added so we can log the service name when there's an error to @Jont828's point below.

File: controllers/azurecluster_reconciler.go (hunk 4 above)
Comment (CecileRobertMichon) [reply controllers/azurecluster_reconciler.go]: this should be solved now, thanks for pointing that out

**Comment by CecileRobertMichon** [comment]
Okay @Jont828 @shysank check out the latest commits. It is a significantly larger refactor but I think it addresses your concerns. I added a new interface called `ServiceReconciler` which is a `Reconciler` that also has `Name()` and `IsManaged()` . The `Name()` func allows us to retrieve the service name for better error logging and for identifying the groups service which avoid assuming ordering. After I did this I realized that we were still assuming ordering by skipping the group delete in the AzureCluster reconciler (loop was going from len-1 to 1 instead of 0). So I added another commit on top that adds `IsManaged` which allows us to check if the group is managed by CAPZ directly from the reconciler. Feel free to tell me this is a scope creep, I kept it unsquashed so we can extract the commit if needed. 

I also included AzureMachinePool and AzureManagedControlPlane reconcilers in the changes since they use the same pattern. 

I think this gets us closer to what we want and we can discuss adding "groups" of services (or priorities, or a dependency graph) if we want to move towards concurrency as a follow-up. The `IsManaged` interface is also something I had wanted to do previously in #1684 to allow us to do caching of the vnet tags and standardize the "is resource managed" story.

File: azure/interfaces.go [hunk 7]
```diff
@@ -24,18 +24,17 @@ import (
...
-// Example: virtualnetworks service would offer Reconcile/Delete methods.
+// Reconciler is a generic interface for a controller reconciler which has Reconcile and Delete methods.
 type Reconciler interface {
 	Reconcile(ctx context.Context) error
 	Delete(ctx context.Context) error
 }
 
-// CredentialGetter is a Service which knows how to retrieve credentials for an Azure
-// resource in a resource group.
-type CredentialGetter interface {
```
Comment (CecileRobertMichon) [inline azure/interfaces.go:36]: this wasn't used anymore

File: azure/services/privatedns/privatedns.go [hunk 8]
```diff
@@ -219,3 +226,9 @@ func (s *Service) isVnetLinkManaged(ctx context.Context, resourceGroupName, zone
 	tags := converters.MapToTags(zone.Tags)
 	return tags.HasOwned(s.Scope.ClusterName()), nil
 }
+
+// IsManaged returns always returns true.
+// TODO: separate private DNS and VNet links so we can implement the IsManaged method for each.
```
Comment (CecileRobertMichon) [inline azure/services/privatedns/privatedns.go:231]: this one is a TODO for after #1715 is merged

File: azure/services/publicips/publicips.go [hunk 9]
```diff
@@ -149,3 +156,8 @@ func (s *Service) isIPManaged(ctx context.Context, ipName string) (bool, error)
 	tags := converters.MapToTags(ip.Tags)
 	return tags.HasOwned(s.Scope.ClusterName()), nil
 }
+
+// IsManaged returns always returns true as public IPs are managed on a one-by-one basis.
```
Comment (CecileRobertMichon) [inline azure/services/publicips/publicips.go:160]: this one is a little awkward. Each public IP can be either managed or unmanaged depending on its own tags and there is no overall "resource type is managed by CAPZ" or not, which makes it different from how the other services behave. Since we're not using this func anywhere (but it's there to satisfy the interface) I opted to always return true since that's the most cautious outcome: we assume public IPs are managed, but then we check on each one individually (https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2146/files#diff-fe943e74189aa531d871d7b774c371755043f5bd1b7e8ab14ad3d2b5fc023301R148). Open to suggestions if anyone has a better idea.

**Comment by CecileRobertMichon** [comment]
looks like e2e tests are failing delete... looking

**Comment by CecileRobertMichon** [comment]
should I rebase or keep the commits separate?

**Comment by shysank** [comment]
> should I squash or keep the commits separate?

I'm fine with keeping the commits as is.

**Comment by Jont828** [comment]
I like the interface changes!
/lgtm
